src/cfn/drift.ts

```diff
@@ -0,0 +1,91 @@
+import * as aws from 'aws-sdk';
+import * as cli from 'cli-color';
+import * as _ from 'lodash';
+import {sprintf} from 'sprintf-js';
+import calcElapsedSeconds from '../calcElapsedSeconds';
+import {GenericCLIArguments} from '../cli/utils';
+import getCurrentAWSRegion from '../getCurrentAWSRegion';
+import mkSpinner from '../spinner';
+import {SUCCESS} from '../statusCodes';
+import timeout from '../timeout';
+import * as yaml from '../yaml';
+import {formatLogicalId, formatSectionHeading, calcPadding} from './formatting';
+import getReliableStartTime from './getReliableStartTime';
+import {getStackDescription} from './getStackDescription';
+import {getStackNameFromArgsAndConfigureAWS} from "./getStackNameFromArgsAndConfigureAWS";
+import {summarizeStackDefinition} from "./summarizeStackDefinition";
+
+
+async function getAllStackResourceDrifts(StackName: string) {
+  const cfn = new aws.CloudFormation();
+  let res = await cfn.describeStackResourceDrifts({StackName}).promise();
+  let drifts = res.StackResourceDrifts
+  while (!_.isUndefined(res.NextToken)) {
+    res = await cfn.describeStackResourceDrifts({StackName, NextToken: res.NextToken}).promise();
+    drifts = drifts.concat(res.StackResourceDrifts);
+  }
+  return drifts.filter(d => d.StackResourceDriftStatus != 'IN_SYNC');
+}
+
+export async function updateStackDriftData(stack: aws.CloudFormation.Stack, cachePeriodSeconds: number = 60*5) {
+  const cfn = new aws.CloudFormation();
+  const startTime = await getReliableStartTime();
+  const recheckIfBefore = new Date(startTime.getTime() - (1000 * cachePeriodSeconds /* ms */));
+  if (stack.DriftInformation &&
+    (stack.DriftInformation.StackDriftStatus == 'NOT_CHECKED'
+      || stack.DriftInformation.LastCheckTimestamp! < recheckIfBefore)) {
+    console.log();
+    const spinnerStartTime = new Date();
+    const spinnerText = "Checking for stack drift.";
+    const spinner = mkSpinner(cli.xterm(240)(spinnerText));
+    const {StackDriftDetectionId} = await cfn.detectStackDrift({StackName: stack.StackName}).promise();
+    while (true) {
+      const statusResponse = await cfn.describeStackDriftDetectionStatus({StackDriftDetectionId}).promise();
+      spinner.stop();
+      if (statusResponse.DetectionStatus !== "DETECTION_IN_PROGRESS") {
+        break;
+      } else {
+        spinner.start();
```

Comment:
Is there a reason that `spinner` `start`/`stop` need to be called in this way? It seems to me `start` could be called before the loop and `stop` in the "then" branch of this conditional (or after the `while` loop).

Reply:
Cut and paste from another loop elsewhere.